File: Week 5 -- 1D spatial models/Homework/hw5.R
library(TMB)
library(data.table)
library(ggplot2)

rm(list=ls())
set.seed(5)
setwd("/Users/chacalle/Documents/classes/2018_FSH556/Week 5 -- 1D spatial models/Homework/")
source("Homework_Week_5_simulator.R")
compile("hw5.cpp")
dyn.load(dynlib("hw5"))

run_simulation_experiment <- function(include_spatial_trend = T) {

  # generate simulated data
  sim_data <- data.table(Sim_Fn())
  setnames(sim_data, c("s_i", "a_i", "l_i", "true_linf_i")) # to be consistent with hw description
  sim_data <- sim_data[order(s_i)] # order data by latitude

  # Build inputs
  Data = list(s_i = sim_data$s_i, a_i = sim_data$a_i, l_i = sim_data$l_i)
  Parameters = list(log_l_0 = log(1), log_k = log(0.1), log_sigma2_measure = 1,
                    log_rho = log(0.1), log_sigma2_spatial = 1,
                    log_beta_0 = log(20), beta_s = 0,
                    epsilon_i = rep(0, nrow(sim_data)))
  Random = c("epsilon_i")

  # optionally include spatial trend
  Map <- list()
  if (!include_spatial_trend) {
    Map[["beta_s"]] <- factor(NA)
  }

  # Build object
  Obj = MakeADFun(data = Data, parameters = Parameters, random = Random, map = Map, DLL="hw5")

  # Obj$fn( Obj$par)
  # Obj$gr( Obj$par )
  # Obj$fn( Obj$env$last.par)
  # Obj$gr( Obj$par )

  # Optimize
  Opt = TMBhelper::Optimize(obj=Obj)

  # calculate rmse for asymptotic maximum size
  Report = Obj$report()
  sim_data[, pred_linf_i := Report$l_inf_i]
  Report$l_inf_i <- NULL
  Report$rmse <- sim_data[, sqrt(mean((true_linf_i - pred_linf_i) ^ 2))]
  setDT(Report)

  Report[, include_spatial_trend := include_spatial_trend]

  return(Report)
}

simulation_results <- lapply(1:10, function(i) {
  results <- lapply(c(F,T), function(v) {
    run_simulation_experiment(v)
  })
  results <- rbindlist(results)
  results[, experiment := i]
})
simulation_results <- rbindlist(simulation_results)
simulation_results <- melt(simulation_results, id.vars = c("include_spatial_trend", "experiment"))
mean_results <- simulation_results[, list(mean = mean(value)), by = c("include_spatial_trend", "variable")]

p <- ggplot(simulation_results, aes(x = value, fill = include_spatial_trend)) +
  geom_density(alpha = 0.5) +
  geom_vline(data = mean_results, aes(xintercept = mean, colour = include_spatial_trend)) +
  facet_wrap( ~ variable, scales = "free") +
  theme_bw()
print(p)

File: Week 5 -- 1D spatial models/Homework/hw5.cpp

#include <TMB.hpp>

// dlnorm
template<class Type>
Type dlognorm(Type x, Type meanlog, Type sdlog, int give_log=0){
  //return 1/(sqrt(2*M_PI)*sd)*exp(-.5*pow((x-mean)/sd,2));
  Type logres = dnorm( log(x), meanlog, sdlog, true) - log(x);
  if(give_log) return logres; else return exp(logres);
}

template<class Type>
Type objective_function<Type>::operator() () {

  // Data
  DATA_VECTOR(s_i);
  DATA_VECTOR(a_i);
  DATA_VECTOR(l_i);

  // Parameters
  PARAMETER(log_l_0);
  PARAMETER(log_k);
  PARAMETER(log_sigma2_measure);
  PARAMETER(log_rho);
  PARAMETER(log_sigma2_spatial);
  PARAMETER(log_beta_0);
  PARAMETER(beta_s);

  // Random effects
  PARAMETER_VECTOR(epsilon_i);

  // Objective funcction
  int n_i = s_i.size();
  vector<Type> jnll_comp(2);
  jnll_comp.setZero();

  Type l_0 = exp(log_l_0);
  Type k = exp(log_k);
  Type sigma2_measure = exp(log_sigma2_measure);
  Type rho = exp(log_rho);
  Type sigma2_spatial = exp(log_sigma2_spatial);
  Type beta_0 = exp(log_beta_0);

  // Probability of random effects
  vector<Type> dist_i(n_i);
  jnll_comp(1) -= dnorm(epsilon_i(0), Type(0), pow(sigma2_spatial, 0.5), true);
  for (int i=1; i < n_i; i++) {
    dist_i(i) = s_i(i) - s_i(i - 1);
    jnll_comp(1) -= dnorm(epsilon_i(i), pow(rho, dist_i(i)) * epsilon_i(i - 1), pow(sigma2_spatial*(1 - pow(rho, 2 * dist_i(i))), 0.5), true);
  }

  // Probability of data conditional on random effects
  vector<Type> l_inf_i(n_i);
  vector<Type> l_i_hat(n_i);
  for (int i = 0; i < n_i; i++) {
    l_inf_i(i) = beta_0 * exp(beta_s * s_i(i)) * exp(epsilon_i(i)); // this is exactly like the simulator function
    l_i_hat(i) = l_inf_i(i) - ((l_inf_i(i) - l_0) * exp(-k * a_i(i)));
    jnll_comp(0) -= dlognorm(l_i(i), log(l_i_hat(i)), pow(sigma2_measure, 0.5), true);
  }

  // Reporting
  Type jnll = jnll_comp.sum();

  REPORT(l_inf_i);
  REPORT(l_0);
  REPORT(k);
  REPORT(sigma2_measure);
  REPORT(rho);
  REPORT(sigma2_spatial);
  REPORT(beta_0);
  REPORT(beta_s);

  return jnll;
}
